File: alarmclock/src/main/java/com/externalalarmclock/alarmclock/ChannelFactory.java
package com.externalalarmclock.alarmclock;

import com.externalalarmclock.lib.rpiws281x.ERpiWs281xStripType;
import com.externalalarmclock.lib.rpiws281x.RpiWs281xChannel;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonTypeName;

@JsonTypeName("channel")
public class ChannelFactory implements IChannelFactory {
    @JsonProperty
	private int ledCount = -1;
    @JsonProperty
	private int gpioNum = -1;
    @JsonProperty
	private byte brightness = (byte) 0xFF;
    @JsonProperty
	private ERpiWs281xStripType stripType = ERpiWs281xStripType.SK6812_STRIP_GRBW;

	@Override
	public RpiWs281xChannel createChannel() {
		RpiWs281xChannel channel = new RpiWs281xChannel();

		channel.setBrightness(brightness);
		channel.setGpioNum(gpioNum);
		channel.setLedCount(ledCount);
		channel.setStripType(stripType);
		
		return channel;
	}
}

File: alarmclock/src/main/kotlin/com/externalalarmclock/alarmclock/ChannelFactory.kt
package com.externalalarmclock.alarmclock

import com.externalalarmclock.lib.rpiws281x.ERpiWs281xStripType
import com.externalalarmclock.lib.rpiws281x.RpiWs281xChannel
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.annotation.JsonTypeName
import io.dropwizard.jackson.Discoverable

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
internal interface IChannelFactory : Discoverable {
    fun createChannel(): RpiWs281xChannel
}

@JsonTypeName("channel")
class ChannelFactory : IChannelFactory {
    @JsonProperty
    private val ledCount = -1
    @JsonProperty
    private val gpioNum = -1
    @JsonProperty
    private val brightness = 0xFF.toByte()
    @JsonProperty
    private val stripType = ERpiWs281xStripType.SK6812_STRIP_GRBW

    override fun createChannel(): RpiWs281xChannel {
        val channel = RpiWs281xChannel()

        channel.brightness = brightness
        channel.gpioNum = gpioNum
        channel.ledCount = ledCount
        channel.stripType = stripType

        return channel
    }
}